### ISSUE TITLE
Support for Filter method .contains().

### ISSUE THREAD

**Issue by hopecee** [datanucleus/datanucleus-neo4j#37]
Hi!
We already have the following Filter methods,
`.eq()   => equals to`  
`.ne()  => not  equal to`
`.startsWith()  => beginning with`
`.matches() => exact value`

can we have method to search for string which contain a string value?
For instance, if a string like **`Appletogether`**, searching for **`toget`** is true.

I have not seen the method for that.

**Comment by andyjefferson** [comment]
If you want a feature adding you need to fork the code and contribute the feature via a GitHub pull request.

**Comment by hopecee** [comment]
I need to learn that now. You can put me through.

**Comment by andyjefferson** [comment]
FYI Valid JDOQL methods are shown on this link http://www.datanucleus.org:15080/products/accessplatform_5_1/jdo/query.html#jdoql_methods
No idea why you would need to fork datanucleus-core, because that then affects ALL datastores. The only possible addition to datanucleus-core would be to add the METHOD "contains" for String values, since JAVA has a method String.contains(...).

**Comment by andyjefferson** [comment]
The JDOQL is 
stringField.contains(someValue)

which maps on to a generic compilation of InvokeExpression. InvokeExpression is what you should process in the neo4j plugin. i.e in QueryToCypherMapper.java

**Comment by hopecee** [comment]
Sorry, Please how do I get that?

**Comment by andyjefferson** [comment]
processInvokeExpression in that class I just quoted! just look at other plugins that do querying, and they all handle methods

**Comment by hopecee** [comment]
I am having small issue on how to release on GitHub. The Zip file doesn't look like the normal jar file.

**Comment by andyjefferson** [comment]
No idea what that means. To contribute on GitHub you provide a PULL REQUEST, via the Pull Requests option on the project.

**Comment by hopecee** [comment]
I will make  PULL REQUEST, but wont it be to test the commit before making PULL REQUEST?

**Comment by andyjefferson** [comment]
It is for YOU to test the proposed changes, before submitting the PULL REQUEST. Hence it is in your local codebase, so you test it locally

**Comment by hopecee** [comment]
Yes I test it locally. Shouldn't I extract the jar and run it on my IDE? Or is there any way to test it on GitHub?

**Comment by andyjefferson** [comment]
GitHub is a repository. It doesn't test anything. You test it. There are some simply tests under the repository "tests", navigating to "jdo/neo4j", as per what the documentation says.

Also, to correct your earlier statement, there is no "startsWith" or "matches" supported by this plugin, as you can see from the processInvokeExpression method. No methods are currently supported.

**Comment by hopecee** [comment]
I followed another example before you come up with that.
This what I did but I have not been able to test it.
[8d93f4a](https://github.com/hopecee/datanucleus-core/commit/8d93f4ae81f2f0d9a1aaf5254431d77baaebf8b6)
[d7a59ac](https://github.com/hopecee/datanucleus-core/commit/d7a59acf7f53e3189f1a2a0bea16ede30c572e9c)
[738ac85](https://github.com/hopecee/datanucleus-core/commit/738ac857674572a48fddf7b921b73cd3bd2c1bd3)
[cd9abf9](https://github.com/hopecee/datanucleus-core/commit/cd9abf97befe02f7dc11c5024f84968b626b2fa3)

[f01cb9f](https://github.com/hopecee/datanucleus-neo4j/commit/f01cb9fcb70eb99e500542ecc2cab00bdebb6f05)
[32b506d](https://github.com/hopecee/datanucleus-neo4j/commit/32b506da1802a08d06814ea185aa3106dd8e0d68)

Is this another way of doing it?

**Comment by andyjefferson** [comment]
JDOQL/JPQL do NOT have any OPERATOR called "CONTAINS". "Contains" is a METHOD. You should not be going anywhere near datanucleus-core. JDOQL allows METHODS to be defined as per what Java allows, and as I've already said, the equivalent JDOQL that you are proposing here is 
`stringField.contains(someValue)`

which is converted into a generic compilation of 
`InvokeExpression{[PrimaryExpression{this.stringField}].contains({arg})}`

**Comment by hopecee** [comment]
Ok. I got you.

**Comment by hopecee** [comment]
Following your footsteps, I have done this [c3b6fb6](https://github.com/hopecee/datanucleus-neo4j/commit/c3b6fb64e6b13752144526ed8c61ffce8d1b4472). My problem is that I have not been able to see where those changes reflected.

**Comment by andyjefferson** [comment]
That change makes no sense to me. You have a JDOQL query of 

> SELECT FROM mydomain.Person WHERE this.name.contains('so')

mapping on to Cypher clause of 

> contains(this.name)

I'm pretty sure there is no Cypher function for Strings called "contains" http://neo4j.com/docs/developer-manual/current/cypher/functions/#header-query-functions-string
so what is the Cypher supposed to be?! After all you said this was possible.

**Comment by hopecee** [comment]
I thought we could get something like this, not that I was sure;

    List<Product> prods = null;

    JDOPersistenceManager jdopm = (JDOPersistenceManager) pm;
    QProduct cand = QProduct.candidate();
    JDOQLTypedQuery<Product> tq = jdopm.newJDOQLTypedQuery(Product.class);

    prods = tq.filter(cand.contains(value)).executeList();

**Comment by hopecee** [comment]
Since that method is not available, I used this code;

    Query<Product> tq2 = jdopm.newQuery("SELECT  FROM " + Product.class.getName() + " WHERE :name.contains("+value+")");

I got this stack trace;

    [filter:InvokeExpression{[ParameterExpression{name}].contains(VariableExpression{amp})}]
    [symbols: amp type=unknown, this type=com.hopecee.tublex.neo4j.jdo.model.Product, name type=unknown]
    2018-01-14 20:11:24.759 [qtp69673470-83] Query - JDOQL Query : Compile (datastore) of "SELECT   FROM com.hopecee.tublex.neo4j.jdo.model.Product WHERE :name.contains(amp)"
    [INFO ] 2018-01-14 20:11:24.759 [qtp69673470-83] General - >> exception in filter
    org.datanucleus.exceptions.NucleusException: Parameter ParameterExpression{name} is not currently set, so cannot complete the compilation

**Comment by hopecee** [comment]
It seems that this String method is not  supported yet.  They could only support `.contains()` in [Collection Methods](http://www.datanucleus.org/products/accessplatform_3_2/jdo/jdoql.html).

Is there any way to do this?

**Comment by andyjefferson** [comment]
Only you know wtf 'value' is so nothing to say; you get a variableexpression so must be inputting something weird. You cannot use jdoqltyped queries with string.contains ... Obviously since there is no method in the jdo spec, hence why i post string-based jdoql. Jdoql method invocation is present in 'tests' project jdo/neo4j and all work for me.

**Comment by hopecee** [comment]
Ok. I have list of names for the Product name, like "Apple", "Man", "AppleMan", "Orange"
I am trying to search for Product with name that contains variable `"apple"` .

**Comment by hopecee** [comment]
I have seen it. https://github.com/datanucleus/tests/blob/master/jdo/neo4j/src/test/org/datanucleus/test/JDOQLTest.java. 
Is this the best way to go? Is it not too cumbersome?

**Comment by andyjefferson** [comment]
Cumbersome? It's a TEST. Any issue for adding a feature needs a test. Should be the first thing you do.

As for the overall "I am trying to search for Product with a name containing", JDOQL has "matches" allowing regexp. You could implement that. Or you could implement "contains".

**Comment by hopecee** [comment]
`JDOQL has "matches" allowing regexp`. Thanks for This. It works perfectly well. I tried it before, I did not know it could be used for regexp. I used string value directly on it before but it did not work as I wanted.

These code works;

    tq.filter(can.matches("(?i).*"+value+".*")).executeList(); //case-insensitive

Thanks again.

**Comment by hopecee** [comment]
`JDOQL has "matches" allowing regexp`. Thanks for This. It works perfectly well. I tried it before, I did not know it could be used for regexp. I used string value directly on it before but it did not work as I wanted.

These code works;

    String value = "appl";
    tq.filter(cand.matches("(?i).*"+value+".*")).executeList(); // case-insensitive


Thanks again.

**Comment by andyjefferson** [comment]
As already said, that is NOT implemented to execute IN THE DATASTORE; it executes in-memory! You have to implement it to work as a method in the datastore

**Comment by hopecee** [comment]
Please give an example on how to use the methods you add to QueryToCypherMapper. Let me see what is going on.

**Comment by andyjefferson** [comment]
I already have, IN THE TESTS. That does `toUpperCase`. These examples are the same as you can find in the DN docs, or JDO spec

**Comment by hopecee** [comment]
Look at what I have come up with here,
[7eca91a](https://github.com/hopecee/datanucleus-neo4j/commit/7eca91a46206f6e07b12278c101363b7e2f6b3be)
[b38fd46](https://github.com/hopecee/datanucleus-neo4j/commit/b38fd463f2b176ec9a70edc16c8854e587088b54)

    Query q1 = jdopm.newQuery("SELECT FROM " + Product.class.getName() + "  WHERE this.name.contains('apple')");
    prods = (List<Product>) q1.execute();

Is the execution still in in-memory?

**Comment by andyjefferson** [comment]
An Operator is an INTERNAL JDOQL/JPQL Operator. There is NO "=~" operator in JDOQL/JPQL and so there should not be here either. You should simply generate the Cypher query you think that JDOQL maps on to. Can't get simpler

Whether it is in-memory or in-datastore is for YOU to see ... because the LOG tells you!

**Comment by hopecee** [comment]
It gives me `Invalid compilation` when I use `Neo4jStringExpression()` or `Neo4jLiteral()`.

     String propName = invokedFieldExpr.getFieldName();
     String value = literalValue.getLiteral().toString();
     String cypherText = propName + " =~ '(?i).*" + value + ".*'";
     Neo4jExpression neo4jExpr = new Neo4jStringExpression(cypherText);
     System.out.println("invokedFiecypherText.neo4jExpr : " + neo4jExpr.getCypherText());
     stack.push(neo4jExpr);
     return neo4jExpr;

That is my reason for using `Neo4jBooleanExpression()`. I am yet to find out why  `Neo4jStringExpression()` gives  `Invalid compilation`. Does it only meant for cypher function?

**Comment by andyjefferson** [comment]
Why would you try to use Neo4jStringExpression????? It would be mindless. As JRE javadocs tell you, String.contains(CharSequence) returns a Boolean. Hence Neo4jBooleanExpression is what you use!! 

If you can't work it out from that then frankly it is your problem. I have other things to do.

**Comment by hopecee** [comment]
I am here to learn. This my commit works. Is anything wrong with it?
[7eca91a](https://github.com/hopecee/datanucleus-neo4j/commit/7eca91a46206f6e07b12278c101363b7e2f6b3be)
[b38fd46](https://github.com/hopecee/datanucleus-neo4j/commit/b38fd463f2b176ec9a70edc16c8854e587088b54)

    Query q1 = jdopm.newQuery("SELECT FROM " + Product.class.getName() + "  WHERE this.name.contains('apple')");
    prods = (List<Product>) q1.execute();

**Comment by andyjefferson** [comment]
Anything wrong with it ? I have already told you on multiple occasions that you should not be touching OPERATIONs, since those are internal JDOQL/JPQL operations and there is no such "=~" in JDOQL/JPQL. So remove that!! You simply need to form the Cypher query equivalent for "String.contains(arg)" and return it in a Neo4jBooleanExpression ...

**Comment by hopecee** [comment]
I did not understand you before. My eyes did not go toward that constructor;

    public Neo4jBooleanExpression(String cypher) {
        cypherText = cypher;
    }

I have done these;
[11d6b58](https://github.com/hopecee/datanucleus-neo4j/commit/11d6b5809958ce628f0417b15f0578ffddfe062b)
[ffe9e6d](https://github.com/hopecee/datanucleus-neo4j/commit/ffe9e6dd79b96b508ce9045b46c3391d25b17c33)

It is still working as expected, but there is one issue with Literal.
Literal with less than two letters does not work here.

**Comment by andyjefferson** [comment]
You are ignoring "neo4jExprArgs" which is the processed form of the "args" from earlier in that method; use those and not "args".
You should check that the literal *value* is of type String (or Character).
You should check if the argument is a parameter, and use the parameter value.
Can Neo4j Cypher allow the argument to be a property? i.e `WHERE p.name =~ '(?i).*' + p.otherName + '.*'` ? in which case you need to support the arg being a PrimaryExpression.
You need to provide a test or tests that demonstrate this (to the "tests" project jdo/neo4j).

You are ignoring DataNucleus code conventions http://www.datanucleus.org/documentation/development.html#coding_standards

**Comment by hopecee** [comment]
I do not think it can allow the argument to be a Parameter unless  the parameter name is converted to  String.
You know, `=~` is use to match regular expressions which is in String form.

    =~ 'regexp'
The documentation is here, http://neo4j.com/docs/developer-manual/current/cypher/clauses/where/#_regular_expressions

**Comment by hopecee** [comment]
Added Support for String.matches(), To except REGEXP 
[fd3756f](https://github.com/hopecee/datanucleus-neo4j/commit/fd3756f71d33060af2bdfcdaf76b3ea182b27277)

**Comment by andyjefferson** [comment]
I don't see anything here that isn't in the current codebase
